File: src/environment.py
"""World geometry: wall penalty field and apple placement (CLAUDE.md invariants 1 & 6).

Responsibilities:
- Maintain the list of live apples, all guaranteed to spawn with their full body
  inside the safe zone (dist_to_wall >= penalty_zone.width + apple.radius).
- Expose the gradient wall-penalty formula for use by agents (Phase 5).
- Handle deferred respawn: an eaten apple leaves the live list for
  ``apple.respawn_delay`` ticks (CDC §5.2), then reappears at a fresh safe
  position. Energy stays apple-equivalent per tick (invariant n°2).
- No eating, no agent state, no rendering.
"""

from __future__ import annotations

from random import Random

import numpy as np

from src.apple import Apple
from src.config import AppleConfig, PenaltyZoneConfig, SimConfig, WorldConfig


class Environment:
    """World model: geometry, penalty field, and apple population."""

    def __init__(self, config: SimConfig, rng: Random) -> None:
        self._world: WorldConfig = config.world
        self._pz: PenaltyZoneConfig = config.penalty_zone
        self._apple_cfg: AppleConfig = config.apple

        zw = self._pz.width
        r = self._apple_cfg.radius
        self._x_min = zw + r
        self._x_max = self._world.width - zw - r
        self._y_min = zw + r
        self._y_max = self._world.height - zw - r

        if self._x_max <= self._x_min or self._y_max <= self._y_min:
            raise ValueError(
                "penalty_zone.width + apple.radius leaves no safe spawn region; "
                "reduce penalty_zone.width or apple.radius"
            )

        self.apples: list[Apple] = [
            Apple(*self._random_safe_position(rng))
            for _ in range(self._apple_cfg.count)
        ]
        # Apples eaten this period, counting down to their respawn.
        self._pending: list[Apple] = []
        # Cached (x, y) NumPy arrays of ``apples``, used by Agent's vectorised
        # raycast (perf-critical: called once per living agent per tick).
        # Invalidated (set to None) whenever ``apples`` changes membership or an
        # apple's position moves (mark_eaten / tick_respawns), then rebuilt lazily
        # on the next access — apples change at most a couple of times per tick
        # (~1 eaten, ~1 respawned) versus hundreds of agents reading them.
        self._apple_coords_cache: tuple[np.ndarray, np.ndarray] | None = None

    # ------------------------------------------------------------------ #
    # Public geometry helpers
    # ------------------------------------------------------------------ #

    def dist_to_wall(self, x: float, y: float) -> float:
        """Minimum distance from (x, y) to any wall."""
        return min(x, self._world.width - x, y, self._world.height - y)

    def penalty_at(self, x: float, y: float) -> float:
        """Energy drained per tick at (x, y) — gradient, not binary (invariant n°2).

        penalty = max_drain * (1 - dist / zone_width)  if dist < zone_width
                = 0.0                                   otherwise
        """
        dist = self.dist_to_wall(x, y)
        if dist < self._pz.width:
            return self._pz.max_drain * (1.0 - dist / self._pz.width)
        return 0.0

    def in_safe_zone(self, x: float, y: float) -> bool:
        """True when (x, y) is outside the penalty zone (agent-facing helper)."""
        return self.dist_to_wall(x, y) >= self._pz.width

    def live_apple_coords(self) -> tuple[np.ndarray, np.ndarray]:
        """Cached ``(x, y)`` NumPy arrays of every live apple, one entry each.

        Rebuilt lazily the first time it's read after ``apples`` changes; reused
        as-is by every agent that senses in between (see cache comment above).
        """
        if self._apple_coords_cache is None:
            xs = np.fromiter(
                (a.x for a in self.apples), dtype=np.float64, count=len(self.apples)
            )
            ys = np.fromiter(
                (a.y for a in self.apples), dtype=np.float64, count=len(self.apples)
            )
            self._apple_coords_cache = (xs, ys)
        return self._apple_coords_cache

    def mark_eaten(self, apple: Apple) -> None:
        """Remove a just-eaten apple from play; it will respawn after the delay.

        The apple leaves ``apples`` for ``_pending`` and starts its countdown.
        Its position is left untouched until ``tick_respawns`` relocates it.
        """
        self.apples.remove(apple)
        apple.respawn_timer = self._apple_cfg.respawn_delay
        self._pending.append(apple)
        self._apple_coords_cache = None

    def tick_respawns(self, rng: Random, tick: int = 0) -> None:
        """Advance every pending apple's countdown; respawn the ones that are due.

        Decrement each pending timer; when it reaches 0 the apple is relocated to
        a fresh safe position and returned to ``apples``. Called once per tick
        (Phase 6, step 8). Logs nothing — rendering is Phase 7. ``tick`` stamps
        the respawned apple's ``spawn_tick`` (diagnostics-only, see
        ``src.diagnostics``); the default 0 keeps every call site that doesn't
        care about it (most tests) unchanged.
        """
        still_pending: list[Apple] = []
        for apple in self._pending:
            apple.respawn_timer -= 1
            if apple.respawn_timer <= 0:
                apple.respawn_timer = 0
                apple.x, apple.y = self._random_safe_position(rng)
                apple.spawn_tick = tick
                self.apples.append(apple)
                self._apple_coords_cache = None
            else:
                still_pending.append(apple)
        self._pending = still_pending

    # ------------------------------------------------------------------ #
    # Private helpers
    # ------------------------------------------------------------------ #

    def _random_safe_position(self, rng: Random) -> tuple[float, float]:
        """Uniform draw inside the radius-inset safe rectangle (invariant n°6)."""
        return (
            rng.uniform(self._x_min, self._x_max),
            rng.uniform(self._y_min, self._y_max),
        )

File: src/lib/alife/environment.ts
// ═══════════════════════════════════════════════════════════════════
//  environment.ts — World geometry, penalty zone, apple management
//
//  Base: Python environment.py
//  Invariant n°6: apples spawn fully inside the safe zone.
//  Deferred respawn: eaten apple leaves for respawn_delay ticks.
// ═══════════════════════════════════════════════════════════════════

import type { SimConfig, AppleConfig } from './config';
import type { SeededRNG } from './genome';

export class Apple {
  x: number;
  y: number;
  respawnTimer: number;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
    this.respawnTimer = 0;
  }
}

export class Environment {
  apples: Apple[] = [];
  private _pending: Apple[] = [];

  private readonly _worldWidth: number;
  private readonly _worldHeight: number;
  private readonly _penaltyWidth: number;
  private readonly _penaltyMaxDrain: number;
  private readonly _appleRadius: number;
  private readonly _appleEnergy: number;
  private readonly _respawnDelay: number;

  // Safe spawn bounds (inset by penalty zone + apple radius)
  private readonly _xMin: number;
  private readonly _xMax: number;
  private readonly _yMin: number;
  private readonly _yMax: number;

  get worldWidth(): number { return this._worldWidth; }
  get worldHeight(): number { return this._worldHeight; }
  get penaltyWidth(): number { return this._penaltyWidth; }
  get penaltyMaxDrain(): number { return this._penaltyMaxDrain; }
  get appleRadius(): number { return this._appleRadius; }
  get appleEnergy(): number { return this._appleEnergy; }

  constructor(config: SimConfig, rng: SeededRNG) {
    const w = config.world;
    const pz = config.penalty_zone;
    const ac = config.apple;

    this._worldWidth = w.width;
    this._worldHeight = w.height;
    this._penaltyWidth = pz.width;
    this._penaltyMaxDrain = pz.max_drain;
    this._appleRadius = ac.radius;
    this._appleEnergy = ac.energy;
    this._respawnDelay = ac.respawn_delay;

    this._xMin = pz.width + ac.radius;
    this._xMax = w.width - pz.width - ac.radius;
    this._yMin = pz.width + ac.radius;
    this._yMax = w.height - pz.width - ac.radius;

    if (this._xMax <= this._xMin || this._yMax <= this._yMin) {
      throw new Error('penalty_zone.width + apple.radius leaves no safe spawn region');
    }

    for (let i = 0; i < ac.count; i++) {
      this.apples.push(new Apple(...this._randomSafePosition(rng)));
    }
  }

  // ── Geometry helpers ─────────────────────────────────────

  distToWall(x: number, y: number): number {
    return Math.min(x, this._worldWidth - x, y, this._worldHeight - y);
  }

  /** Gradient wall penalty (per tick) — invariant n°2 */
  penaltyAt(x: number, y: number): number {
    const dist = this.distToWall(x, y);
    if (dist < this._penaltyWidth) {
      return this._penaltyMaxDrain * (1.0 - dist / this._penaltyWidth);
    }
    return 0.0;
  }

  // ── Apple lifecycle ──────────────────────────────────────

  markEaten(apple: Apple): void {
    const idx = this.apples.indexOf(apple);
    if (idx >= 0) {
      this.apples.splice(idx, 1);
    }
    apple.respawnTimer = this._respawnDelay;
    this._pending.push(apple);
  }

  tickRespawns(rng: SeededRNG): void {
    const stillPending: Apple[] = [];
    for (const apple of this._pending) {
      apple.respawnTimer--;
      if (apple.respawnTimer <= 0) {
        apple.respawnTimer = 0;
        const [nx, ny] = this._randomSafePosition(rng);
        apple.x = nx;
        apple.y = ny;
        this.apples.push(apple);
      } else {
        stillPending.push(apple);
      }
    }
    this._pending = stillPending;
  }

  // ── Private ──────────────────────────────────────────────

  private _randomSafePosition(rng: SeededRNG): [number, number] {
    return [
      rng.uniform(this._xMin, this._xMax),
      rng.uniform(this._yMin, this._yMax),
    ];
  }
}

// ── Geometry pure functions (shared by agent raycasts) ─────

export { clamp, rayAngles } from './geometry';

export function rayWalls(
  px: number, py: number, dx: number, dy: number,
  width: number, height: number,
): number | null {
  let best: number | null = null;
  if (dx > 0) best = closer(best, (width - px) / dx);
  else if (dx < 0) best = closer(best, -px / dx);
  if (dy > 0) best = closer(best, (height - py) / dy);
  else if (dy < 0) best = closer(best, -py / dy);
  return best;
}

export function rayCircle(
  px: number, py: number, dx: number, dy: number,
  cx: number, cy: number, radius: number,
): number | null {
  const fx = px - cx;
  const fy = py - cy;
  const b = 2.0 * (fx * dx + fy * dy);
  const c = fx * fx + fy * fy - radius * radius;
  const disc = b * b - 4.0 * c;
  if (disc < 0) return null;
  const sqrtDisc = Math.sqrt(disc);
  const tNear = (-b - sqrtDisc) / 2.0;
  if (tNear > 0) return tNear;
  const tFar = (-b + sqrtDisc) / 2.0;
  if (tFar > 0) return tFar;
  return null;
}

function closer(current: number | null, candidate: number): number | null {
  if (candidate <= 0) return current;
  if (current === null || candidate < current) return candidate;
  return current;
}